### ISSUE TITLE
Is transaction rollback called automatically if node process shutdown while sql is executed??

### ISSUE THREAD

**Issue by tanimoto-true** [brianc/node-postgres#1458]
Transaction begins, sql is executed and error occurs before transaction commit or rollback, then node process is shutdown, 
is transaction automatically rollbacked in this case??

node-postgres : v7.3
node : v8.1.4

**Comment by abenhamdine** [comment]
No, there's no particular transaction management features in this library because it's intented to be low level.
So if you don't fire a ROLLBACK statement yourself, the transaction will be pending.
See https://node-postgres.com/features/transactions

**Comment by boromisp** [comment]
I believe the transaction will be rolled back as soon as postgres realizes that the connection has been lost.
You should see similar messages in the postgresql log:

```
could not receive data from client: Connection reset by peer
unexpected EOF on client connection with an open transaction
```

If the client has been shut down "cleanly" then it could be immediately.

If the client simply just disappeared (crash, network config, etc.), then it would depend on the client OS, server OS and postgresql configurations, if and when the database detects the dropped connection.

https://www.postgresql.org/docs/current/static/runtime-config-connection.html#GUC-TCP-KEEPALIVES-IDLE

https://www.gnugk.org/keepalive.html


Otherwise, if the 'commit' has been sent just before shutdown, then this would apply: https://www.postgresql.org/docs/current/static/protocol-flow.html#AEN113117.

> If the query is outside any transaction block (BEGIN ... COMMIT sequence) then its results might be committed before the disconnection is recognized.

In short: undefined.

**Comment by vitaly-t** [comment]
On any server an unclosed transaction becomes locked, or more specifically, all tables involved in the transaction become locked.

You can read about PostgreSQL locks here: http://blog.nordeus.com/dev-ops/postgresql-locking-revealed.htm

**Comment by tanimoto-true** [comment]
Thank you guys for your help.

As a conclusion, this package doesn't manage unexpected shutdown while transaction runs.
So, maybe using ```try - catch - finally```  for every transaction code and manage errors by my end is the best option as doc says!! 

Anyway, Thanks a lot!!
